docs/design/2020-06-24-placement-rules-in-sql.md

```diff
@@ -0,0 +1,391 @@
+
+- Author(s):     [djshow832](https://github.com/djshow832) (Ming Zhang)
+- Last updated:  2020-06-24
+- Discussion at: https://docs.google.com/document/d/18Kdhi90dv33muF9k_VAIccNLeGf-DdQyUc8JlWF9Gok
+
+## Movitation
+
+TiDB supports placement rules, which can define the placement of data in a more flexible and more granular way. But it only provides configuration files to define them, and it’s complicated.
+
+This article proposes an approach to configure placement rules through statements. TiDB server parses the statements and notify PD to perform the change. In this way, usability can be improved.
+
+## Define placement rules
+
+There are 2 kinds of operations on the placement:
+
+* ADD: Add new `count` replicas to the table according to `label` and `role` configuration.
+* ALTER: Override the current replica configuration. Old configuration is discarded.
+
+They’re both achieved by executing `ALTER TABLE` statements.
+
+### Adding replicas
+
+Adding new replicas can be done by one or more `ADD PLACEMENT` clauses:
+
+```sql
+ALTER TABLE table_name
+	ADD PLACEMENT LABEL=label ROLE=role COUNT=count,
+	...
+```
+
+This statement indicates TiDB to add replicas for all data, including indexes, of table `table_name`.
+
+`ADD PLACEMENT` is just a part of alter options, just like `ADD COLUMN` or `ADD CONSTRAINT`.
+
+To define multiple roles at once, Multiple `ADD PLACEMENT` clauses can appear in a single `ALTER TABLE` statement, each for one Raft role:
+
+```sql
+ALTER TABLE table_name
+	ADD PLACEMENT LABEL="+zone=sh" ROLE=leader COUNT=1,
+	ADD PLACEMENT LABEL="-zone=sh" ROLE=follower COUNT=2;
+```
+
+This statement indicates PD to schedule the leader to `sh`, and do not schedule followers to `sh`.
+
+`ADD PLACEMENT` can be used to separate data on different data centers, which may help to reduce multi-region latency.
+
+Adding placement can be done through adding a placement rule in PD. The statement must wait until the PD returns a message.
+
+### Altering placement
+
+Altering current placement rules can be done by one or more `ALTER PLACEMENT` clauses:
+
+```sql
+ALTER TABLE table_name
+	ALTER PLACEMENT LABEL=label ROLE=role COUNT=count,
+	...
+```
+
+This statement indicates TiDB to overwrite the current placement rule with the same `role`. It affects all data, including indexes, of table `table_name`.
+
+This kind of operation can be used in adding more replicas for important data, or adding a TiFlash replica.
+
+Altering placement can be done through override the current placement rule by a new one.
+
+Note that when the rules in the statement conflict, an error should occur. For example, an error should be reported when executing this statement:
+
+```sql
+ALTER TABLE test
+	ALTER PLACEMENT LABEL="+zone=sh" ROLE=voter COUNT=3,
+	ALTER PLACEMENT LABEL="+zone=bj" ROLE=voter COUNT=3;
+```
+
+`+zone=sh` conflicts with `+zone=bj`, because they’re defining the same role `voter`. These two constraints can not be satisfied both.
```

Comment:
why it's not able to add voters in both bj and sh?

Reply:
Good catch. I've updated the doc.
`ADD` will combine the rules with the same role, but `ALTER` will replace the old rule with the same role.
It's different from CRDB. CRDB has only one rule for each table, and only the preferences of voters and leaders are defined. TiDB is more complicated because it support voters, leaders, followers and learners. So we must separate the roles to different rules.